File: AtCoder/ABC056/D_.py
import sys
input = sys.stdin.readline

N, K = list(map(int, input().split()))
A = list(map(int, input().split()))

# K-a <= x <= K-1

from time import time

t = time()

dp_L = [[False]*K for _ in range(N)]
dp_L[0][0] = True
for i, a in enumerate(A[:-1]):
    dp_L_next = dp_L[i+1]
    for j, d in enumerate(dp_L[i]):
        dp_L_next[j] = d
    if a>=K: continue
    for j, d in enumerate(dp_L[i][:-a]):
        dp_L_next[j+a] |= d

dp_R = [[False]*K for _ in range(N)]
dp_R[0][0] = True
for i, a in enumerate(A[:0:-1]):
    dp_R_next = dp_R[i+1]
    for j, d in enumerate(dp_R[i]):
        dp_R_next[j] = d
    if a>=K: continue
    for j, d in enumerate(dp_R[i][:-a]):
        dp_R_next[j+a] |= d


print(time()-t)
# for i, d in enumerate(dp_L):
#     print(i,d)
# 
# for i, d in enumerate(dp_R):
#     print(i, d)

t = time()

ans = 0
for i, a in enumerate(A):
    dl = dp_L[i]
    dr = dp_R[-(i+1)]
    
    noneed = True
    ml = -K
    for j in range(K):
        fl = dl[j]
        fr = dr[K-1-j]
        if fl: ml = j
        if fr and j-ml<a:
            noneed = False
            break
    if noneed: ans+=1

print(ans)

print(time()-t)

File: AtCoder/ABC056/D.cpp
#include<iostream>
#include<vector>
using namespace std;
#define ll long long

int main(){
  ll N, K;
  cin >> N >> K;
  vector<ll> A(N);
  for(ll i=0; i<N; ++i) cin >> A[i];

  vector<bool> tmp(K, false);
  vector<vector<bool>> dp_L(N, tmp), dp_R(N, tmp);
  dp_L[0][0] = dp_R[0][0] = true;

  for(ll i=0; i<N-1; ++i){
    dp_L[i+1] = dp_L[i];
    dp_R[i+1] = dp_R[i];
    ll aL = A[i], aR = A[N-1-i];
    if(aL<K){ for(ll j=0; j<K-aL; ++j) dp_L[i+1][j+aL] = dp_L[i][j]; }
    if(aR<K){ for(ll j=0; j<K-aR; ++j) dp_R[i+1][j+aR] = dp_R[i][j]; }
  }

  ll ans = 0;
  for(ll i=0; i<N; ++i) {
    vector<bool>& dl = dp_L[i];
    vector<bool>& dr = dp_R[N-1-i];

    bool noneed = true;
    ll ml = -K;
    for(ll j=0; j<K; ++j) {
      if (dl[j]) { ml = j; }
      if (dr[K-1-j] && j-ml<A[i]) {
        noneed = false;
        break;
      }
    }
    if (noneed) ++ans;
  }

  cout << ans << endl;

  return 0;
}
